lib/ash/subject.ex

```diff
@@ -0,0 +1,539 @@
+defmodule Ash.Subject do
+  @moduledoc """
+  Provides a consistent API for common operations across `Ash.Changeset`,
+  `Ash.Query`, and `Ash.ActionInput`. It allows you to write generic code that works
+  with any of these types without needing to pattern match or special-case your logic.
+  """
+
+  @type t :: Ash.Changeset.t() | Ash.Query.t() | Ash.ActionInput.t()
+
+  @doc """
+  Adds an error or list of errors to the subject.
+
+  Supports all subject types (Changeset, Query, ActionInput) and maintains
+  type consistency.
+
+  ## Parameters
+
+    * `subject` - The subject to add errors to
+    * `errors` - Error or list of errors to add
+    * `path` - Path where the error occurred (optional, defaults to [])
+  """
+  @spec add_error(t(), Ash.Error.error_subject() | list(Ash.Error.error_subject())) :: t()
+  @spec add_error(
+          t(),
+          Ash.Error.error_subject() | list(Ash.Error.error_subject()),
+          Ash.Error.path_subject()
+        ) :: t()
+  def add_error(subject, errors, path \\ [])
+
+  def add_error(subject, [], _path), do: subject
+
+  def add_error(%Ash.Changeset{} = subject, errors, path) when is_list(errors) do
```

Comment:
I think we can simplify this interface by *not* accepting path here. You can just `Ash.Error.set_path` on the error before calling these.